smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/diff/EndpointSigV4Migration.java

```diff
@@ -0,0 +1,249 @@
+import software.amazon.smithy.rulesengine.aws.language.functions.EndpointAuthUtils;
+import software.amazon.smithy.rulesengine.language.Endpoint;
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait;
+import software.amazon.smithy.utils.Pair;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Emit diff validation for SigV4 migration in the {@code @smithy.rules#endpointRuleSet} trait.
+ *
+ * Specifically, SigV4 ({@code aws.auth#sigv4}) to SigV4A ({@code aws.auth#sigv4a}) due to a subset of credentials
+ * usable with SigV4 that are not migratable to SigV4A.
+ *
+ * @see <a href="https://smithy.io/2.0/aws/aws-auth.html">AWS Authentication Traits</a>
+ * @see <a href="https://smithy.io/2.0/additional-specs/rules-engine/specification.html#endpoint-authschemes-list-property">Endpoint {@code authSchemes} list property</a>
+ */
+@SmithyInternalApi
+public final class EndpointSigV4Migration extends AbstractDiffEvaluator {
+    @Override
+    public List<ValidationEvent> evaluate(Differences differences) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        Model oldModel = differences.getOldModel();
+        ServiceIndex oldServiceIndex = ServiceIndex.of(oldModel);
+        Model newModel = differences.getNewModel();
+        ServiceIndex newServiceIndex = ServiceIndex.of(newModel);
+
+        // Validate Service effective auth schemes
+        List<ChangedShape<ServiceShape>> serviceChanges = differences
+            .changedShapes(ServiceShape.class)
+            .collect(Collectors.toList());
+        for (ChangedShape<ServiceShape> change : serviceChanges) {
+            ServiceShape oldServiceShape = change.getOldShape();
+            ServiceShape newServiceShape = change.getNewShape();
+
+            if (!oldServiceShape.hasTrait(EndpointRuleSetTrait.ID)
+                || !newServiceShape.hasTrait(EndpointRuleSetTrait.ID)) {
+                continue;
+            }
+
+            Optional<Pair<EndpointRuleSetTrait, EndpointRuleSetTrait>> endpointRuleSetOpt =
+                change.getChangedTrait(EndpointRuleSetTrait.class);
+            Optional<Pair<AuthTrait, AuthTrait>> authOpt =
+                change.getChangedTrait(AuthTrait.class);
+            List<String> oldModeledAuthSchemes = getModeledAuthSchemes(oldServiceIndex, oldServiceShape);
+            List<String> newModeledAuthSchemes = getModeledAuthSchemes(newServiceIndex, newServiceShape);
+            // Validate diffs for changes to `@smithy.rules#endpointRuleSet` and `@auth` and effective auth schemes
+            if (!endpointRuleSetOpt.isPresent()
+                && !authOpt.isPresent()
+                && oldModeledAuthSchemes.equals(newModeledAuthSchemes)) {
+                continue;
+            }
+
+            EndpointRuleSetTrait oldErs = endpointRuleSetOpt
+                .map(Pair::getLeft)
+                .orElse(oldServiceShape.expectTrait(EndpointRuleSetTrait.class));
```

Comment:
can just always get oldServiceShape.expectTrait.

Reply:
Changed to just expect `EndpointRuleSetTrait` directly.